Transport Parameters are not APLN Specific (#3973)
quicwg/base-drafts: draft-ietf-quic-transport.md

Endpoints use transport parameters to authenticate the negotiation of
connection IDs during the handshake; see {{cid-auth}}.

When sent by the client, transport parameters are not specific to a particular
ALPN being negotiated.  Therefore, a client cannot have application protocol
specific values in the transport parameters.  Application protocols should take
this into account when designing versioning schemes, and not place restrictions
on the values in the transport parameters that might prevent protocol changes

> in future version of the application protocol.

Reviewer: Regarding the last sentence, how about saying something like: _Application protocols SHOULD NOT introduce prohibitive requirements that prevent other application protocols from working._

I am not sure if this problem can be scoped to different versions of one protocol. While that could be the case for HTTP, there could be cases where a client might ask for using different protocols that provide the same service. To give an example, a client might send send an ALPN of ("h3" (for doing DNS over HTTP), "dns" (for doing DNS over QUIC)) to a DNS server.

Author: @kazuho I'm not sure how you'd go about not making _prohibitive requirements_ for other protocols without understanding the set of all other possible protocols **or** by setting no requirements at all.

Reviewer: @nibanks If that's unclear, saying something like the following might be better: "should not introduce restrictions to the transport parameters, as doing so might prevent clients from offering multiple application protocols."

I think that addresses the two examples that you laid out in the original comment of #3965.

Reviewer 2: There is a difference between requirements and recommendations.  We do recommend in HTTP/3 that a certain number of unidirectional stream be available, but it is [just a recommendation](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-6.2-4).

I think that this could be clearer then:

> Protocol negotiation with Application Layer Protocol Negotiation (ALPN; see {{?ALPN=RFC7301}}) allows clients to offer multiple application protocols during connection establishing. The single set of transport parameters offered by a client apply to all application protocols that the client offers.  Application protocols can make recommendations about the values for transport parameters, such as suggesting minimum values for stream or flow control limits.  Application protocols that place hard constraints on transport parameters could make it impossible for clients to attempt a connection that includes other application protocols with different constraints.

Reviewer 3: I'd like to make some editorial suggestions to martin's proposal, but I like that text.

Endpoints use transport parameters to authenticate the negotiation of
connection IDs during the handshake; see {{cid-auth}}.

Application Layer Protocol Negotiation (ALPN; see {{?ALPN=RFC7301}}) allows
clients to offer multiple application protocols during connection
establishment. The transport parameters that a client includes during the
handshake apply to all application protocols that the client offers. Application
protocols can recommend values for transport parameters, such as the initial

> flow control limits. However, application protocols that require values for

Reviewer: ```suggestion
flow control limits. However, application protocols that require specific values for
```

I think we are talking about the requirement of values rather than if TP take values.

Reviewer 2: Or
```suggestion
flow control limits. However, application protocols that constrain values for
```

Reviewer 2: Or

```suggestion
flow control limits. However, application protocols that set constraints on values for
```

> transport parameters will make it impossible for a client to offer multiple

Reviewer 2: ```suggestion
transport parameters could make it impossible for a client to offer multiple
```

> application protocols with conflicting requirements.

Reviewer 2: ```suggestion
application protocols if these requirements conflict.
```

Endpoints use transport parameters to authenticate the negotiation of
connection IDs during the handshake; see {{cid-auth}}.

> Protocol negotiation with Application Layer Protocol Negotiation (ALPN; see
> {{?ALPN=RFC7301}}) allows clients to offer multiple application protocols with
> connection attempts. The single set of transport parameters offered by a client
> apply to all application protocols that the client offers. Application protocols
> can make recommendations about the values for transport parameters, such as
> suggesting minimum values for stream or flow control limits. Application
> protocols that place hard constraints on transport parameters could make it
> impossible for clients to attempt a connection that includes other application
> protocols with different constraints.

Reviewer 3: ```suggestion
Application Layer Protocol Negotiation (ALPN; see {{?ALPN=RFC7301}}) allows
clients to offer multiple application protocols during connection
establishment. The transport parameters that a client includes during the
handshake apply to all application protocols that the client offers. Application
protocols can recommend values for transport parameters, such as the initial
flow control limits. However, application protocols that require values for
transport parameters will make	it impossible for a client to offer multiple
application protocols with conflicting requirements.
```
